### PR TITLE
[FLINK-6241] [table] codeGen dataStream aggregates that use ProcessFunction

### PR DESCRIPTION
**Pull request by shaoxuan-wang**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -239,6 +241,212 @@ class CodeGenerator(
...
+    * @param clazz      AggregateHelper Function to be generated.
+    * @param aggFields  position (in the input Row) of the input value for each aggregate
+    * @param aggregates List of all aggregate functions
+    * @param generator  code generator instance
+    * @param inputType  Input row type
+    *
+    * @tparam F Flink Function to be generated.
+    * @return instance of GeneratedFunction
+    */
+  def generateAggregateHelper[F <: Function](
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: rename to `generateAggregations`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -239,6 +241,212 @@ class CodeGenerator(
...
+      name: String,
+      clazz: Class[F],
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: I don't think we need the `Class` parameter

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala (hunk 1 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: The generic type can be removed

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/generated.scala [hunk 3]
```diff
@@ -58,6 +58,16 @@ case class GeneratedFunction[F <: Function, T <: Any](
   code: String)
 
 /**
+  * Describes a generated aggregate helper function
+  *
+  * @param name class name of the generated Function.
+  * @param code code of the generated Function.
+  */
+case class AggregateHelperFunction(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/generated.scala]: rename to `GeneratedAggregationsFunction`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.api.common.functions.Function
+import org.apache.flink.types.Row
+
+/**
+  * Base class for Aggregate Helper Function.
+  */
+abstract class AggregateHelper extends Function {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala]: rename to `GeneratedAggregations`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+    * Calculate the results from accumulators, and set the results to the output
+    *
+    * @param accumulators the accumulators (saved in a row) which contains the current
+    *                     aggregated results
+    * @param output       output results collected in a row
+    * @param rowOffset    offset of the position (in the output row) where the accumulators
+    *                     starts
+    */
+  def setOutput(accumulators: Row, output: Row, rowOffset: Int)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala]: rename to `setAggregationResults()`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    */
+  def createAccumulator(): Row
+
+  /**
+    * Init the accumulators, and save them to the input accumulators Row.
+    *
+    * @param input  input values bundled in a row
+    * @param output output results collected in a row
+    */
+  def forwardValueToOutput(input: Row, output: Row)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala]: rename to `setForwardedFields()`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    * @param input        input values bundled in a row
+    */
+  def retract(accumulators: Row, input: Row)
+
+  /**
+    * Init the accumulators, and save them to a accumulators Row.
+    *
+    * @return a row of accumulators which contains the aggregated results
+    */
+  def createAccumulator(): Row
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateHelper.scala]: rename to `createAccumulators()` (+s)?

**Comment by shaoxuan-wang** [comment]
Thanks @fhueske, these changes look good to me.

**Pull request closed without merging**